### PR TITLE
Parallel run with one proc w/o partition file

### PR DESCRIPTION
**Pull request by stcui007**
This PR address issue #703 in ngen framework, i.e., to run a MPI job using one processor without providing a partition file. For more than one processor, you must provide a partition file. If you insist on providing the partition file in the case of one processor, you are still ok.

## Additions

- Define a dummy class `PartitionOne` for to present a whole-hydrofabric partition for a single process, and associated test

## Removals

None

## Changes

- Use `PartitionOne` when running on a single MPI process, rather than expecting a partition file argument on the command line

## Testing

Passed all local test except RoutingPyBindTest.TestRoutingPyBind
Passing all automated testing

## Notes

Command line example to run a job without/with partition file:

    mpirun -n 1 ./cmake_build_mpi/ngen data/catchment_data.geojson '' data/nexus_data.geojson '' data/example_bmi_multi_realization_config.json
 
    mpirun -n 3 ./cmake_build_mpi/ngen data/catchment_data.geojson '' data/nexus_data.geojson '' data/example_bmi_multi_realization_config.json ./test_partition_3.json

## Checklist

- [x] PR has an informative and human-readable title
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code can be automatically merged (no conflicts)
- [x] Code follows project standards (link if applicable)
- [x] Passes all existing automated tests
- [x] Any _change_ in functionality is tested
- [X] New functions are documented (with a description, list of inputs, and expected output)
- [X] Placeholder code is flagged / future todos are captured in comments
- [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [X] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### CODE REVIEW

File: include/core/Partition_Data.hpp
Comment (PhilMiller) [inline include/core/Partition_Data.hpp:7]:
```suggestion
struct PartitionData
{
    using Tuple = std::tuple<int, std::string, std::string, std::string>;
```
Don't declare a global type named `Tuple`

File: include/core/Partition_One.hpp
Comment (PhilMiller) [inline include/core/Partition_One.hpp:49]: These members seem to duplicate the members of `PartitionData`. Why not just use them?

File: include/core/Partition_Data.hpp
Comment (stcui007) [reply include/core/Partition_Data.hpp:7]: I tried to make the change. But it caused some build errors in ```Partition_Parser.hpp```. Apparently, ```Partition_Parser.hpp``` uses this declaration in a few places in its code region. In its current form, the ```Partition_Parser.hpp``` would be the same as its original form in master branch.

File: include/core/Partition_One.hpp
Comment (stcui007) [reply include/core/Partition_One.hpp:49]: These variables are used this way following the example of ```Partition_Parser.hpp```. In this form, it seems the logic in the loop above is simpler. We could use the struct members, but we would seem to use longer variable names with somewhat overuse of the "." operator.

File: include/core/Partition_Data.hpp
Comment (PhilMiller) [reply include/core/Partition_Data.hpp:7]: It would become `PartitionData::Tuple`.

File: include/core/Partition_One.hpp
Comment (PhilMiller) [reply include/core/Partition_One.hpp:49]: Syntactic simplicity isn't a good reason to duplicate data in memory and add extra assignments, as well as just adding extra declarations that anyone reading the code later has to make sense of.

File: include/core/Partition_Data.hpp
Comment (stcui007) [reply include/core/Partition_Data.hpp]: Right. Implemented.

File: include/core/Partition_One.hpp
Comment (stcui007) [reply include/core/Partition_One.hpp]: Revised as requested.

File: include/core/Partition_Data.hpp
Comment (PhilMiller) [inline include/core/Partition_Data.hpp:8]:
```suggestion
    using Tuple = std::tuple<int, std::string, std::string, std::string>;
```

File: include/core/Partition_Data.hpp
Comment (PhilMiller) [inline include/core/Partition_Data.hpp:5]:
```suggestion
```

File: include/core/Partition_One.hpp
Comment (PhilMiller) [inline include/core/Partition_One.hpp:28]:
```suggestion
```

File: include/core/Partition_One.hpp
Comment (PhilMiller) [inline include/core/Partition_One.hpp:41]:
```suggestion
```
```suggestion
        //PartitionData is a struct
```

File: include/core/Partition_Parser.hpp
Comment (PhilMiller) [inline include/core/Partition_Parser.hpp:55]:
```suggestion
```

**Comment by stcui007** [comment]
Thanks Phil for the review. Will work on it first thing in the morning
Monday.

File: test/CMakeLists.txt
Comment (PhilMiller) [inline test/CMakeLists.txt:472]: I'm concerned about making the entire `test_unit` build dependent on having SQLite support enabled.

File: include/core/Partition_One.hpp
Comment (PhilMiller) [inline include/core/Partition_One.hpp:39]:
```suggestion
```
There are no virtual functions here, and nothing at all for a destructor to even do, so we don't need to write one out at all.

File: include/core/Partition_One.hpp
Comment (PhilMiller) [inline include/core/Partition_One.hpp:20]:
```suggestion
```
The compiler will provide a default constructor without any need to write something out. Specifying a non-default constructor can create unintended side effects.

File: test/CMakeLists.txt
Comment (PhilMiller) [reply test/CMakeLists.txt:472]: Maybe the test case could say

```
         #if !defined(NGEN_WITH_SQLITE3)
            GTEST_SKIP() << "Compiled without SQLite support"
         #endif
```
?

File: test/utils/Partition_One_Test.cpp
Comment (PhilMiller) [inline test/utils/Partition_One_Test.cpp:129]:
```suggestion
    for( auto& id: duplicates){
        // Add printing code as needed
    }

    ASSERT_EQ(duplicates.size(), 0);
```
Unit tests should never be calling `exit()` directly, since that kills the whole test process, and doesn't properly report the failed test result.

The loop over duplicates doesn't need to be conditional, since if there are none, it does nothing.

Same applies to 1b below.

**Comment by PhilMiller** [comment]
I pushed fixes for some of the smaller things I noted in my review. I wanted to call my concerns about the test code to your attention.

**Comment by stcui007** [comment]
Could you be more explicit?

**Comment by stcui007** [comment]
Ignore my previous question if you don't have any more to add. Just read your comments on the test codes.

File: test/CMakeLists.txt
Comment (stcui007) [reply test/CMakeLists.txt:472]: The problem is with the ```geopackage``` is involved in the test code.

**Comment by stcui007** [comment]
I just put a couple of comments in the PR page.

File: test/CMakeLists.txt
Comment (program--) [reply test/CMakeLists.txt:472]: Haven't reviewed this, but IIRC from an earlier discussion, you could take out the GeoPackage reading, and just use GeoJSON, since it isn't needed to test the functionality of this PR.

**Comment by stcui007** [comment]
I have made an update by making Partition_One an independent test in
analogy with some other packages.

File: .github/workflows/documentation.yml
Comment (PhilMiller) [inline .github/workflows/documentation.yml:49]: Why and how did this step get duplicated?

File: test/utils/Partition_One_Test.cpp
Comment (PhilMiller) [inline test/utils/Partition_One_Test.cpp:56]: Nothing ever sets `catchmentDataFile`, so this condition will never be true. Why does the test depend on Geopackage at all? If it's not being used, just delete it!

File: include/core/Partition_One.hpp
Comment (PhilMiller) [inline include/core/Partition_One.hpp:4]: The guard for `NGEN_MPI_ACTIVE` isn't necessary here. MPI's not actually referenced or used by any of this code. Removing the guard would simplify test code by removing the weird `#define NGEN_MPI_ACTIVE`

File: .github/workflows/documentation.yml
Comment (stcui007) [reply .github/workflows/documentation.yml:49]: No idea. The only way that I can think of is during rebasing. At no time that I even touched that file. Did rebasing senses something different even human eyes see they are exactly the same? Any way, I will remove the repeat.

File: test/utils/Partition_One_Test.cpp
Comment (stcui007) [reply test/utils/Partition_One_Test.cpp:56]: Agree, If we need to use gpkg in the future, we can add it back with some revision.

File: include/core/Partition_One.hpp
Comment (stcui007) [reply include/core/Partition_One.hpp:4]: The presence of ```NGEN_MPI_ACTIVE``` does make the code look a bit cumbersome. However, Partition_One is specifically designed for this case. The source codes are under this condition. On the other hand, I assume if the test code works w/o MPI. it is probably ok.

Comment (stcui007) [reply include/core/Partition_One.hpp:4]: The ```NGEN_MPI_ACTIVE``` is needed as Partition_One.hpp needs it. Without it, the test code won't compile. The Parttion_Test.cpp using Partition_Parser.hpp works in a similar fashion. So this is kept in place.

File: test/utils/Partition_One_Test.cpp
Comment (PhilMiller) [inline test/utils/Partition_One_Test.cpp:117]:
```suggestion
    for (auto& id: duplicates) {
        std::cout << "duplicates string set contains " << id << std::endl;
    }
```
The `if` condition is testing for `duplicates` containing an element that is an empty string (`id` is a `std::string`. That makes no sense. If `duplicates` contains any elements at all, the error message applies, and it should print the element(s) found.

The exact same concern applies to test 1b below.

Comment (stcui007) [reply test/utils/Partition_One_Test.cpp:117]: The if condition tests for duplicates that's not empty by the ```negation operator```. So if that condition holds, we print out a message that says there are duplicates. Then break out of the loop so we don't need to print out a lot of lines if there are many in the duplicates.

File: test/utils/Partition_One_Test.cpp
Comment (PhilMiller) [inline test/utils/Partition_One_Test.cpp:95]: I just realized an oversight in these tests - we also need to check that `catchment_ids` and `nexus_ids` contain the right number of elements.

Also, I just noticed that the comparison isn't the partition output against the original `catchment_collection`, but just against itself. The test as a whole should affirm that the set in the single partition is identical to the set in the input it was given.

**Comment by stcui007** [comment]
Thanks for the careful review. Will work on it.

**Comment by PhilMiller** [comment]
Code looks good. I'm going to wait until we've got #767 merged to rebase and merge this.

**Review by PhilMiller — APPROVED** [review]

**Pull request merged**
